### PR TITLE
Service indicator checks and tests for KDF, DRBG and ECDH

### PR DESCRIPTION
**Pull request by samuel40791765**
### Issues:
Addresses `CryptoAlg-835`

### Description of changes: 
Service indicator checks and tests for:
* KDF:
  * `CRYPTO_tls1_prf` approved for TLS 1.0/MD5_SHA-1, TLS 1.2/SHA2-256, TLS 1.2/SHA2-384, and TLS 1.2/SHA2-512
* DRBG:
  * `CTR_DRBG_generate` 
  * Test for `RAND_bytes` which is the only external way for a consumer to indirectly interact with the DRBG functions.
  * Service indicator locks within `crypto/fipsmodule/rsa/padding.c` because unapproved functions within it call `RAND_bytes` and `EVP_Digest`
* ECDH:
  * `ECDH_compute_key_fips`
  
### Call-outs:
* Added new parameter under static function for `CRYPTO_tls1_prf` to help differentiate md5/sha1 from using only md5 and only sha1.
* Service indicator locks within `crypto/fipsmodule/rsa/padding.c` because unapproved functions within it call `RAND_bytes` and `EVP_Digest`
 
### Testing:
New tests in `service_indicator_test.cc` testing dimension.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: crypto/fipsmodule/rsa/padding.c [hunk 1]
```diff
@@ -325,44 +339,43 @@ int RSA_padding_add_PKCS1_OAEP_mgf1(uint8_t *to, size_t to_len,
   }
 
   size_t mdlen = EVP_MD_size(md);
+  size_t emlen = to_len - 1;
```
Comment (torben-hansen) [inline crypto/fipsmodule/rsa/padding.c:342]: Why move this?

File: crypto/fipsmodule/service_indicator/service_indicator_test.cc [hunk 2]
```diff
@@ -440,6 +547,117 @@ TEST(ServiceIndicatorTest, AESGCM) {
   ASSERT_EQ(approved, AWSLC_NOT_APPROVED);
 }
 
+TEST(ServiceIndicatorTest, ECDH) {
+  int approved = AWSLC_NOT_APPROVED;
+
+  bssl::UniquePtr<EC_GROUP> group(EC_GROUP_new_by_curve_name(NID_secp224r1));
```
Comment (torben-hansen) [inline crypto/fipsmodule/service_indicator/service_indicator_test.cc:553]: What about the other groups, and groups not approved?

File: crypto/fipsmodule/service_indicator/service_indicator_test.cc [hunk 3]
```diff
@@ -440,6 +547,117 @@ TEST(ServiceIndicatorTest, AESGCM) {
   ASSERT_EQ(approved, AWSLC_NOT_APPROVED);
 }
 
+TEST(ServiceIndicatorTest, ECDH) {
```
Comment (torben-hansen) [inline crypto/fipsmodule/service_indicator/service_indicator_test.cc:550]: Do we need KATs here really? We should be able to generate two ephemeral key pairs and perform an ECDH and run through all curves.

File: crypto/fipsmodule/service_indicator/service_indicator_test.cc [hunk 4]
```diff
@@ -440,6 +547,117 @@ TEST(ServiceIndicatorTest, AESGCM) {
...
+  ASSERT_TRUE(check_test(kDRBGReseedOutput, output, sizeof(kDRBGReseedOutput),"DRBG Reseed KAT"));
+
+  // Test approval of |RAND_bytes|, which is the only way for the consumer to
+  // indirectly interact with the DRBG functions.
+  CALL_SERVICE_AND_CHECK_APPROVED(approved, ASSERT_TRUE(RAND_bytes(output, sizeof(output))));
+  ASSERT_EQ(approved, AWSLC_APPROVED);
+  CTR_DRBG_clear(&drbg);
+}
+
+TEST(ServiceIndicatorTest, TLSKDF) {
```
Comment (torben-hansen) [inline crypto/fipsmodule/service_indicator/service_indicator_test.cc:624]: Nit: Bit of code-reuse here.

File: crypto/fipsmodule/service_indicator/service_indicator_test.cc [hunk 5]
```diff
@@ -11,21 +11,23 @@
 #include <openssl/bn.h>
 #include <openssl/cipher.h>
 #include <openssl/cmac.h>
-#include <openssl/des.h>
-#include <openssl/dh.h>
```
Comment (nebeid) [inline crypto/fipsmodule/service_indicator/service_indicator_test.cc:15]: 👍

File: crypto/fipsmodule/rand/ctrdrbg.c [hunk 6]
```diff
@@ -194,6 +194,7 @@ int CTR_DRBG_generate(CTR_DRBG_STATE *drbg, uint8_t *out, size_t out_len,
   }
 
   drbg->reseed_counter++;
+  FIPS_service_indicator_update_state();
```
Comment (nebeid) [inline crypto/fipsmodule/rand/ctrdrbg.c:197]: Would lines like l.169 need locking? isn't it AES-CTR?

File: crypto/fipsmodule/rsa/padding.c [hunk 7]
```diff
@@ -325,44 +339,43 @@ int RSA_padding_add_PKCS1_OAEP_mgf1(uint8_t *to, size_t to_len,
...
   db[emlen - from_len - mdlen - 1] = 0x01;
   OPENSSL_memcpy(db + emlen - from_len - mdlen, from, from_len);
   if (!RAND_bytes(seed, mdlen)) {
-    return 0;
-  }
-
-  uint8_t *dbmask = OPENSSL_malloc(emlen - mdlen);
-  if (dbmask == NULL) {
-    OPENSSL_PUT_ERROR(RSA, ERR_R_MALLOC_FAILURE);
-    return 0;
```
Comment (nebeid) [inline crypto/fipsmodule/rsa/padding.c:362]: Why are these lines deleted?

File: crypto/fipsmodule/tls/kdf.c [hunk 8]
```diff
@@ -143,23 +152,24 @@ int CRYPTO_tls1_prf(const EVP_MD *digest,
...
-                     label_len, seed1, seed1_len, seed2, seed2_len)) {
+                     label_len, seed1, seed1_len, seed2, seed2_len, using_md5_sha1)) {
       return 0;
     }
 
     // Note that, if |secret_len| is odd, the two halves share a byte.
     secret += secret_len - secret_half;
     secret_len = secret_half;
     digest = EVP_sha1();
+    using_md5_sha1 = 1;
```
Comment (nebeid) [inline crypto/fipsmodule/tls/kdf.c:170]: Will MD5-SHA1 be used or just SHA1? I didn't follow the code closely, sorry.

**Review by nebeid — COMMENTED** [review]
I am yet to review the tests.

File: crypto/fipsmodule/service_indicator/service_indicator.c [hunk 9]
```diff
@@ -139,6 +139,50 @@ void HMAC_verify_service_indicator(const EVP_MD *evp_md) {
...
+    FIPS_service_indicator_update_state();
+  }
+  else if(curve == NID_secp521r1) {
+    FIPS_service_indicator_update_state();
+  }
+}
+
+
+void TLSKDF_verify_service_indicator(const EVP_MD *dgst, const int using_md5_sha1) {
+  // HMAC-MD5/HMACSHA1 (both used concurrently) is approved in TLS 1.0/1.1.
```
[the remark this answers is not in the document]
Comment (samuel40791765) [reply crypto/fipsmodule/service_indicator/service_indicator.c:165]: Leaving it like it originally was for now.

File: crypto/fipsmodule/service_indicator/service_indicator_test.cc (hunk 2 above)
Comment (nebeid) [reply crypto/fipsmodule/service_indicator/service_indicator_test.cc]: In another PR, can this test be parameterised to test all curves? Would be a lot of vectors, though and may not be needed. May be in that case, follow Torben's suggestion of making the test self-consistency rather than KATs.

File: crypto/fipsmodule/service_indicator/service_indicator_test.cc (hunk 3 above)
Comment (samuel40791765) [reply crypto/fipsmodule/service_indicator/service_indicator_test.cc]: Didn't know this was doable at the time, before writing tests for ECDSA. Just updated :)

File: crypto/fipsmodule/service_indicator/service_indicator_test.cc (hunk 2 above)
Comment (samuel40791765) [reply crypto/fipsmodule/service_indicator/service_indicator_test.cc]: Groups other than the approved ones fail directly when using `EC_GROUP_new_by_curve_name` to create a `EC_GROUP`.
https://github.com/awslabs/aws-lc/blob/88d124deee4e948a78d8e710d35af5fc18259a82/include/openssl/ec.h#L104-L115

File: crypto/fipsmodule/service_indicator/service_indicator_test.cc [hunk 10]
```diff
@@ -11,21 +11,23 @@
...
 #include <openssl/nid.h>
+#include <openssl/rand.h>
 #include <openssl/rsa.h>
 #include <openssl/sha.h>
 
 #include "../../test/abi_test.h"
 #include "../../test/file_test.h"
 #include "../../test/test_util.h"
+#include "../rand/internal.h"
+#include "../tls/internal.h"
```
Comment (nebeid) [inline crypto/fipsmodule/service_indicator/service_indicator_test.cc:31]: Should `CRYPTO_tls1_prf` be moved to an external header?

File: crypto/fipsmodule/service_indicator/service_indicator_test.cc [hunk 11]
```diff
@@ -440,6 +547,117 @@ TEST(ServiceIndicatorTest, AESGCM) {
...
+
+  // Test approved usages of KDF.
+  uint8_t tls_output[sizeof(kTLSOutput_sha256)];
+  CALL_SERVICE_AND_CHECK_APPROVED(approved,
+            CRYPTO_tls1_prf(EVP_sha256(), tls_output, sizeof(tls_output), kTLSSecret,
+                            sizeof(kTLSSecret), kTLSLabel, sizeof(kTLSLabel),
+                            kTLSSeed1, sizeof(kTLSSeed1), kTLSSeed2,
+                            sizeof(kTLSSeed2)));
+  ASSERT_TRUE(check_test(kTLSOutput_sha256, tls_output, sizeof(kTLSOutput_sha256), "TLS KDF KAT for sha256"));
+  ASSERT_TRUE(approved);
```
Comment (nebeid) [inline crypto/fipsmodule/service_indicator/service_indicator_test.cc]: Here and the subsequent ASSERTs on `approved`, should they not be ASSERT_EQ?

File: crypto/fipsmodule/service_indicator/service_indicator.c [hunk 12]
```diff
@@ -196,18 +196,13 @@ void ECDH_verify_service_indicator(const EC_KEY *ec_key) {
 }
 
 
-void TLSKDF_verify_service_indicator(const EVP_MD *dgst, const int using_md5_sha1) {
+void TLSKDF_verify_service_indicator(const EVP_MD *dgst) {
   // HMAC-MD5/HMACSHA1 (both used concurrently) is approved in TLS 1.0/1.1.
   // HMAC-SHA256, HMAC-SHA384, HMAC-SHA512 is approved for TLS 1.2.
```
Comment (nebeid) [inline crypto/fipsmodule/service_indicator/service_indicator.c:201]:
```suggestion
  // HMAC-SHA256, HMAC-SHA384, HMAC-SHA512 are approved for TLS 1.2.
```

**Review by nebeid — DISMISSED** [review]

File: crypto/fipsmodule/service_indicator/service_indicator_test.cc [hunk 13]
```diff
@@ -580,6 +676,107 @@ TEST_P(HMAC_ServiceIndicatorTest, HMACTest) {
...
+INSTANTIATE_TEST_SUITE_P(All, KDF_ServiceIndicatorTest, testing::ValuesIn(kKDFTestVectors));
+
+TEST_P(KDF_ServiceIndicatorTest, KDF) {
+  const KDFTestVector &kdfTestVector = GetParam();
+
+  int approved = AWSLC_NOT_APPROVED;
+
+  std::vector<uint8_t> tls_output(32);
+  CALL_SERVICE_AND_CHECK_APPROVED(
+      approved, CRYPTO_tls1_prf(kdfTestVector.func(), tls_output.data(), tls_output.size(),
```
Comment (torben-hansen) [inline crypto/fipsmodule/service_indicator/service_indicator_test.cc:771]: `CRYPTO_tls1_prf` can fail.

**Review by nebeid — APPROVED** [review]
